### ISSUE TITLE
Feature request: API to allow setting thread priority for new Isolates

### ISSUE THREAD

**Issue by maks** [dart-lang/sdk#46943]
Following on from discussion in #44228 I wanted to propose an addition to the API for Isolate spawning which would allow setting the thread priority of a thread allocated to the Isolate or new Isolate group.

I'm not familiar with the implementation details of the thread pool and how threads from it are assigned to Isolates, so I'm not sure if there is currently a means to have a specific thread be assigned or tied to a specific Isolate or Isolate group.

My use case for this is I'm trying to use Dart (on Linux) for audio DSP applications (eg, synthezisers, audio fx processing, etc) and it seems the recommended way to do low latency audio output applications is to use a high priority thread to feed the OS's audio api while trying to ensure that thread never blocks. Given its currently not possible to use FFI with async callbacks into Dart Isolates and that most cross-platform audio libraries (libsoundio, jack, miniaudio) use callbacks from their own high priority "RT" thread, my thought was the next best thing could be to be able to provide a high priority thread for a single Dart Isolates group, which would then have at least 1 isolate calling a sync audio (eg ALSA) api via FFI with audio samples generated in other isolates supplied via its SendPort or FFI pointers if the sendport mechanism itself is too slow.

**Comment by maks** [comment]
Actually I just realised that with FFI, if it's the same thread that does FFI that also runs my Dart code in the Isolate that calls the FFI functions, then I could be a bit checky and just use [pthreads api in native code ](http://www.yonch.com/tech/82-linux-thread-priority) to elevate the priority of that thread myself. But this also would be assuming that the same thread keeps getting scheduled to run the code in my worker "high priority" Isolate, which I'm not clear if that would be the case, even if i only had main + 1 other worker  Isolate?

@mkustermann would you have any thoughts on if this could work or is am I just on a fools errand?

**Comment by mkustermann** [comment]
There's a few things to dissect here:

* **Thread pinning**: Is there a need for the Dart code as well as any possible C code it calls via FFI to always run on the same thread for the existence of that isolate. This is sometimes needed if a C library makes use of TLS (thread local storage) - for example: Dart calls C which then sets up a TLS value calls back into Dart which might call back into a C callback (which accesses TLS key).
   If that was implemented one could use FFI to call out to pthread API to change the priority of a thread and it would stick (since the isolate stays on that thread. One unexpected side-effect of this is that any threads spawned from the former (e.g. Dart spawns garbage collection threads) would inherit the priority.

* **Isolate prioritization**: Does a Dart isolate need to run at a specific priority. Firstly it's not always possible to increase a threads priority - it might require special (e.g. root) privileges for the process (e.g. `setpriority()` on linux allows lowering but not increasing priorities by-default). Most likely one would want to run the code anyway - making this prioritization only a best effort (might or might not apply the priority) solution.

* **Low latency**: Running Dart code (together with some FFI calls) might or might not fit your latency requirements. For example the VM can decide to collect garbage: a) young-generation collections can take 20+ ms b) in low-memory situations the VM can decide to perform a compaction (could take 100+ms). So by running Dart code you give up control over latency requirements to some extend already.

* **Async communication**: An isolate can communicate with a C thread asynchronously via ports. So one can decide to have a dedicated C thread, which sets it's thread priority and pings Dart code when it wants to send it some message or wants an answer for some request (in which case it'd need to wait for the answer - which can increase pause times).

**Comment by maks** [comment]
Thanks for looking at this so quickly @mkustermann !

You make **very** good points about low latency and I have thought myself if this is even feasible in Dart. But I got some hope that  this might be possible by comparing to my entry into Dart: Flutter, which has somewhat similar low latency requirements to meet the 16ms deadlines for 60fps, though I guess that they do that by giving the UI thread higher priority via the embedding.

In regards to thread pinning, I think you are spot on and I think this is what this feature request boils down to as with that it would be up to the user to make use of any OS specific APIs to raise thread priority rather than requiring any of that functionality from Dart itself. I guess this would be the minimal amount of change required in Dart to allow for this functionality and leave it up to any user to do whats best of their use case, in my particular case for example, I'm intending to run on "single-purpose" RPI's, so for instance potentially running my Dart process at elevated, even root user level is not necessarily an issue.  

In regards to GC, yes any pauses could well blow out the time budget, but from my reading it seems that even for native code, its recommended *not* to do any allocation on the high priority audio thread (malloc et al) so for my particular use case I'd be following a similar strategy and avoiding runtime memory allocation in my Dart code as much as possible. Avoiding GC (as much as possible  at least) would also I think alleviate issues of a GC thread inheriting the high priority of the thread in the Isolate that spawned it.
 
So I guess should I change the title of this issue to reflect the change in the requested feature?

**Comment by mkustermann** [comment]
> So I guess should I change the title of this issue to reflect the change in the requested feature?

Either that or file a new issue specifically about thread pinning. We can then also merge  https://github.com/dart-lang/sdk/issues/38315 into it - which fundamentally is about the same problem.

**Comment by maks** [comment]
Thanks @mkustermann ! 
I've changed the title of this issue as I think it would help having the original context here along with the explanation in your comment.  I'm still learning how Core Audio on iOS and MacOS works but suspect thread pinning would also be needed there to use the realtime audio render thread.

**Comment by a-siva** [comment]
Work is being done to allow embedders to control spawning of worker threads instead of having the Dart VM do it (see https://github.com/dart-lang/sdk/issues/44228), would that solve the issues raised here (controlling thread priorities, thread pinning etc.).

**Comment by a-siva** [comment]
@maks I saw the comment posted on the other thread stating the use case presented here is not for a custom embedder but rather having this functionality in the regular Dart SDK.

**Comment by maks** [comment]
Thanks @a-siva yes sorry I should have been more specific, the request here is for using the regular SDK, eg for command line applications

**Comment by gaaclarke** [comment]
I also ran into this problem.  My use case is calling into JNI with dart:ffi.  You can't call into JNI from an arbitrary thread, it has to be registered with the JavaVM via `AttachCurrentThread`.

@maks to your original problem in the description I've written a few synths and this is how you want to setup your code, the fact that Dart doesn't execute on a dedicated thread shouldn't be a problem for you.

1) Write your synth in a non-garbage collected language like C/C++.  You don't want garbage collection interrupting your audio callback and as you make your synthesizer more complex you will be starved for more CPU so you'll want a language that is a thin abstraction over the machine code to make it easier to optimize.
1) Create a lockless ring buffer that will be used to send control data to the synthesizer that is running on the audio callback thread, even using mutexs on that thread will potentially cause buffer underruns (audio glitches resulting from the CPU not responding fast enough).
1) Call from dart with dart:ffi to C/C++ code that will register the audio callback and supply it with the synth it will use to fill the audio buffer and the lockless data buffer that will receive control data
1) When you what to change a parameter for the synthesizer you will use dart:ffi to call into C which will place the data on the lockless ring buffer.  It doesn't matter what thread executes this, it only matters that calls to write to it are serial.
1) In the audio callback read from the lockless ring buffer to modify the synth before calculating the buffer.

Here's some pseudo code:
```c
struct Data {
  Synth* synth;
  LocklessRingBuffer* controlBuffer;
}

struct ControlChange {
  int parameter;
  int value;
}

// Called on some high priority audio thread
static int AudioCallback(float* buffer, int length, void* voidData) {
  Data* data = (Data*)voidData;
  while(LocklessRingBufferHasData()) {
    ControlChange cc = LocklessRingBufferPop(data->controlBuffer);
    SynthSetControlValue(data->synth, cc.parameter, cc.value);
  }
  return SynthFillBuffer(data->synth, buffer, length);
}

// Called on some Dart thread via dart:ffi.
Data* StartSynth() {
  Data* data = CreateSynth();
  AudioApiStartAudioCallback(AudioCallback, data);
  return data;
}

// Called on some Dart thread via dart:ffi.
void KeyDown(Data* data, int key) {
  LocklessRingBufferPush(data->controlBuffer, (ControlChange) {
    .parameter = KEY_DOWN,
    .value = key
  });
}
```

**Comment by maks** [comment]
Thanks for taking the time to write such a detailed reply @gaaclarke ! very much appreciated and very nice to meet another synth developer!

The approach you outline is roughly what I found is done by the [flutter sequencer package](https://github.com/mikeperri/flutter_sequencer).

And yes I've gradually come to the realisation that using the method you've outlined above is really the only viable way at the moment with Dart.

I've read the seminal ["time waits for nothing"](http://www.rossbencina.com/code/real-time-audio-programming-101-time-waits-for-nothing) about the need to avoid not just GC but even manual mem alloc to avoid anything that might block. So I had initially thought I could work around Darts use of GC by having a separate Isolate where I didn't do any runtime allocation doing the audio synth and output to the platform audio api, using a lockless ring buffer to transfer data to the Isolate via FFI pre-allocated memory, hence filing this issue (thanks to mraleph via twitter for suggesting to me that trick to avoid any slowness due to using SendPorts). 

But I've since learnt from feedback from @mkustermann that when GC happens it effects *all* Isolates 😞 plus mraleph in another issue (about compiling Dart to Wasm) said that even if I didn't do any explicit object creation at runtime myself, I might trigger such alloc that Dart would do behind the scenes and hence cause GC.

So I've now (sadly) given up on trying to have the audio synth/effects code in Dart which [I was trying to do in this project](https://github.com/maks/dart_synth) and instead switch over to having the WASM synth in that project output its audio data frames to the system audio api (alsa/pulseaudio/jack etc) all in C/Rust/Zig/etc code rather than routing it via a Dart FFI binding to a audiolib.  

Last year I revived and put back on Play the [
Raph Levien's original "msfa" Dx7 fm synth app](https://github.com/maks/droid-synth) so I'm likely to try to use your recommended approach first with its C code, as I have planned to convert it into a Flutter app.

Given the above I don't really have a use case anymore to request the thread pinning feature, but it sounds like there maybe others that would, but if thats not the case, I'm happy for the issue to be closed.

**Comment by dcharkes** [comment]
> * **Thread pinning**: Is there a need for the Dart code as well as any possible C code it calls via FFI to always run on the same thread for the existence of that isolate. This is sometimes needed if a C library makes use of TLS (thread local storage) - for example: Dart calls C which then sets up a TLS value calls back into Dart which might call back into a C callback (which accesses TLS key).

We also run into this with using JNI. For example a thread needs to be attached to the JVM `AttachCurrentThread`, and `JNIEnv*`s are only valid for the thread for which it was obtained. cc @mahesh-hegde

**Comment by antonbashir** [comment]
I also run into thread pinning issue. [#49463](https://github.com/dart-lang/sdk/issues/49463)
My case has not dependencies with JNI but I use native library which uses thread local values and I need pin Isolate to thread.

**Comment by maks** [comment]
While the title of this issue is about thread pinning, my original usecase for opening this issue was in trying to use Dart via FFI for low latency audio and it seems at least on that front, while not a complete solution, the new [`Dart_PerformanceMode_Latency`](https://github.com/dart-lang/sdk/commit/c6a1eb1b61844b2d733f9e2f4c7754f1920325d7) in the embedder API may be of some help there for future use.

**Comment by dcharkes** [comment]
@knopp's use case needs thread local storage and thread priority setting. So that use case does not to be the main thread and integration with the OS message loop.

If I understand @srmncnk's use case correctly, it would also suffice for the VM to promise to not change the thread after it started. (And the native plugins need to be updated to use FFI instead of method channels.) @srmncnk please correct me if I'm wrong. (https://github.com/flutter/flutter/issues/176053 is an alternative solution on the Flutter side.)

**Comment by srmncnk** [comment]
@dcharkes That's correct. Or course I'm advocating for https://github.com/flutter/flutter/issues/176053 as it's likely easiest for every party involved. On the long run, with some heavy lifting, pinning a Dart Isolate to a thread would suffice too.

**Comment by srmncnk** [comment]
Bumping this in the context of the flutter 3.38 release. Any status updated would be much appreciated.

**Comment by mraleph** [comment]
I don't think anybody is working on this already and the work has not been slated for implementation.

FWIW I think the implementation is fairly trivial - there is simply a question of API design, e.g. do we think that this should _just_ be a parameter to `Isolate.spawn{,Uri}` and `Isolate.run`?

```dart
/// If [dedicatedThread] is `true` then code running in the spawned `Isolate` will
/// execute on a newly created dedicated OS thread. This thread will exist
/// as long as the `Isolate` exists.   
external static Future<Isolate> spawn<T>(
    void entryPoint(T message),
    T message, {
    bool paused = false,
    bool errorsAreFatal = true,
    SendPort? onExit,
    SendPort? onError,
    String? debugName,
    @Since("3.11") bool dedicatedThread = false,
  });
``` 

Is this going to address your needs @srmncnk?

In longer run I think https://github.com/dart-lang/sdk/issues/56841 actually offers a more compelling API design, enabling developer to spawn a thread and synchronously enter a specific isolate.

**Comment by srmncnk** [comment]
@mraleph This will absolutely solve my problem! FWIW I'm happy with the parameter as-is. I would advocate for this solution to be implemented if it's fairy trivial, given the regression we got in flutter 3.38.

https://github.com/dart-lang/sdk/issues/56841 sounds exciting, will be watching as well.

**Comment by maks** [comment]
Wow, 4 years since I opened this issue.
@mraleph  for my part I'm not sure that API would work for my original use case as for instance on MacOS I believe CoreAudio requires audio callbacks to happen on a specific audio thread so I think #56841 would be more relevant and even on Linux I think the key part for low latency audio is to be able to have the audio thread enter a specific Isolate synchronously.

**Comment by knopp** [comment]
Being able to pin isolate thread is certainly useful (thread locals, etc), but I'd be quite hesitant to have core audio callback happening on thread that might be paused for GC...

**Comment by maks** [comment]
> Being able to pin isolate thread is certainly useful (thread locals, etc), but I'd be quite hesitant to have core audio callback happening on thread that might be paused for GC...

*sigh* very good point @knopp ! I keep forgetting that Martin told me quite a while ago that GC is across all isolates so yes, just having the functionality from 56841 wouldn't be enough either as for audio I would be preallocating all the required memory but that Isolates thread could still be paused for GC even if its not a thread owned by the Dart VM.

**Comment by srmncnk** [comment]
@maks FWIW regarding audio, we've had great success with Google Oboe via FFI as we needed really low latency. But we're still in trouble for real-time signal processing, still waiting for this feature.

**Comment by maks** [comment]
> [@maks](https://github.com/maks) FWIW regarding audio, we've had great success with Google Oboe via FFI as we needed really low latency. But we're still in trouble for real-time signal processing, still waiting for this feature.

That sounds very interesting indeed @srmncnk  ! Low latency audio is my interest for this too, any chance you could elaborate more on how you used Oboe via Darts FFI ? Is it in this project? https://github.com/srmncnk/flutter_raw_sound

**Comment by srmncnk** [comment]
@maks No, that was a dead end, especially since, as @knopp pointed out, you're exposed to the GC sooner rather than later. Truthfully a lot of that code was ai generated but essentially it was instructed to replace flutter_raw_sound. And it did, with great success.

**Comment by knopp** [comment]
I'm currently playing with multithreaded rendering with skia graphite (dart only, no flutter), the idea is N+1 threads/isolates for N windows (n * graphite recorder + 1 context processing recorder submissions).  It works quite well, since there is no shared state the port model fits nicely, but it would be helpful to have thread pinning for profiling purposes.

**Comment by srmncnk** [comment]
Any chance of this getting in https://github.com/orgs/dart-lang/projects/110 if it's a low hanging fruit?

**Comment by mraleph** [comment]
@srmncnk yep, that's the plan. I am incorporating necessary low-level APIs [here](https://github.com/dart-lang/language/pull/4620).

**Comment by knopp** [comment]
@mraleph, I'm currently building a (fairly experimental) [skia bindings](https://github.com/knopp/skia_dart) and it would be quite beneficial to be able have a pinned thread for raster thread(s), as it they need custom priority setting. 

Another thing that seems problematic right now is native finalizer possibly being invoked from another thread. When unreffing GPU bound `SkSurface` in some situations it needs to be disposed on the thread it was created. This is fairly easy to detect in finalizer, but redirecting the destructor is not. It would be very helpful to have something like

```cpp
// Returns true is successfully scheduled, false otherwise (i.e. isolate shutting down)
bool Dart_PostCallback(Dart_Isolate isolate, void (*callback)(void *), void *data);
```

exposed in the FFI methods. That would make redirecting the surface destructor quite a bit easier. Otherwise i will need to send the method + arg through `NativeCallable.listener` or a port and call it from dart, seems like a bit of overkill.  But also for this to land on the proper thread we'd need thread pinning.

**Comment by srmncnk** [comment]
Any predictions on how far away we are from this? Q3, Q4, next year?

**Comment by mraleph** [comment]
@srmncnk we don't have precise timeline. All necessary functionality is implemented but hidden under flags. e.g. you can look at this [test](https://github.com/dart-lang/sdk/blob/4866bde1c16272f24c9852f2f59e530258c4353c/tests/ffi/threading_runeventloop_test.dart#L72) which spawns a native thread and then pins an isolate to it, but we have not reached consensus on shipping it yet. I would project at minimum one more quarter to iron it out the details given how things are going.

**Comment by srmncnk** [comment]
Thanks for the insight, much appreciated. Looks good, looking forward to it!
